### ISSUE TITLE
Formatting 'chr' and 'start/end' just before calling in CustomServiceReader

### ISSUE THREAD

**Issue by kbruskiewicz** [igvteam/igv.js#1086]
Our use case involves interpreting data called from a REST API into an IGV annotations track. I was happy to find the CustomServiceReader for this purpose – configuring it is simple and straightforward.

But, when substituting in the strings into the URL, I notice two issues
- Chromosome, when substituted in, come with the 'chr' prefix, but the API I am using only takes the chromosome number.
- Coordinates, when substituted in, get transformed into scientific notation, likely because of how Javascript encodes them. But I need them in integer form.

What would be the best way to intercept or reformat the request header before the coordinates are passed to the call? Or, is it possible to extend the CustomServiceReader and add new readers?

**Comment by jrobinso** [comment]
Hi,  

The chromosome string will be whatever the reference fasta you use defines.  If you use a reference with "chr1" naming that will be used, if you use one with "1" that will be used.  Actually its an arbitrary string.   If you want to mix conventions with the current code you will need to translate it in the web service.   

Alternatively we could add a configuration option that supplies a function that takes chr, start, end and produces a string to pass to the webservice.   This would require some coding, but not much.   If you want to give it a shot and submit a PR I will look at it.  Will need a test case.

RE scientific notation,   I haven't seen that.    Could you provide an example?

**Comment by kbruskiewicz** [comment]
> Scientific Notation

Sure, I've highlighted it in yellow.

![scientific_notation](https://user-images.githubusercontent.com/955221/78413252-76ca4600-75cb-11ea-8751-34784182d2b2.PNG)

The value that gets passed in initially also seems quite large – 308 significant figures – and isn't synchronized with the locus value that I initialize the Browser with.

**Comment by jrobinso** [comment]
Ahh yes, if "visibilityWindow" is not set for a track it queries for the entire chromosome, which is indicated by Number.MAX_VALUE for the end.   I should fix that.   In the meantime you can avoid it by setting a visibilityWindow,  set it to the maximum genomic range that you want queried at one time.  For example visibilityWindow: 1000000 would limit the query to 1 mb.

**Comment by jrobinso** [comment]
Expanding the query interval for web services was not intentional, and seems like the wrong thing to do,  I will push a fix soon for that as well.   In the meantime I suggest setting a visibilityWindow.

**Comment by jrobinso** [comment]
The floating point number should be fixed with the latest push.   Also I'm working on an option to allow the "url" property of the source to be a function, taking {chr, start, end} as input and returning the url to be used.   You should be able to strip of "chr" or do anything else needed there.  I pushed a rather complex example for the cBio copy number segment service,  but all you should need is this part of it
```
  source: {
            url: function (options) {
                const chr = options.chr.startsWith("chr") ? options.chr.substring(3) : options.chr;
                return `http://www.cbioportal.org/api/copy-number-segments/fetch?chromosome=${chr}&projection=SUMMARY`;
            },
       ....
```


full example
```
{
        name: "P101MF - query by chromosome",
        type: "seg",
        sourceType: "custom",
        source: {
            url: function (options) {
                const chr = options.chr.startsWith("chr") ? options.chr.substring(3) : options.chr;
                return `http://www.cbioportal.org/api/copy-number-segments/fetch?chromosome=${chr}&projection=SUMMARY`;
            },
            method: "POST",
            contentType: "application/json",
            body: JSON.stringify([{
                sampleId: "PT101MF",
                studyId: "sarc_mskcc"
            }]),
            queryable: true,   // i.e queryable by genomic range (chromosome only in this case)
            isLog: true,
            mappings: {
                chr: "chromosome",
                value: "segmentMean",
                sampleKey: "uniqueSampleKey",
                sample: "sampleId"
            }
        }
    }
```

**Comment by kbruskiewicz** [comment]
Thanks so much for these changes.

I notice that some of these problems were caught just at the end of a release cycle. Would it be possible to bump a 2.5.1 some time in the near future?

**Comment by jrobinso** [comment]
Yes, probably the end of next week.   In the meantime you can develop with the nightly snapshot, builds are here: http://igv.org/web/snapshot/dist/

**Comment by kbruskiewicz** [comment]
Thanks for pointing out the nightly snapshots to me! I look forward to the new release

One last question: can CustomServiceReaders be used in track types other than annotation, e.g. including GWAS tracks?

**Comment by jrobinso** [comment]
I think I've conflated 2 issues, the custom reader in the commit above was meant for #1089.

Yes to GWAS,  you specify the track  "type=gwas".    The features should have chr, start, end, and value at a minimum.   The the GWAS track decription in the wiki for other details.   I think this issue is resolved, will move discussion to #1089
